src/compiler/sys.ts

```diff
@@ -150,12 +150,17 @@ namespace ts {
                 return result.sort();
             }
 
-            function readDirectory(path: string, extension?: string, exclude?: string[]): string[] {
+            function readDirectory(path: string, extension?: string, exclude?: string[], depth?: number): string[] {
                 const result: string[] = [];
                 exclude = map(exclude, s => getCanonicalPath(combinePaths(path, s)));
-                visitDirectory(path);
+                visitDirectory(path, depth);
                 return result;
-                function visitDirectory(path: string) {
+
+                function visitDirectory(path: string, depth?: number) {
+                    if (depth <= 0) {

```

Comment:
This depends on the funky behavior that "undefined <= 0" is false in order not to break when the optional argument isn't provided. Can you check the argument is provided before checking it is 0 or less (e.g. "depth !== undefined && ..").